File: go/euler37.go
// euler37 находит сумму одиннадцати простых чисел, из которых можно
// выбрасывать цифры как справа налево, так и слева направо.
//
// Решение задачи Эйлера №37.
package main

import (
	"fmt"
	"math"
	"time"
)

func eratosfen(number int) []bool {
	sieve := make([]bool, number+1)
	for i := range sieve {
		sieve[i] = true
	}
	sieve[0] = false
	sieve[1] = false
	for i := 2; i*i <= number; i++ {
		if sieve[i] {
			for j := i * i; j <= number; j += i {
				sieve[j] = false
			}
		}
	}
	return sieve
}

func isPrime(number int) bool {
	if number < 2 {
		return false
	}
	if number == 2 || number == 3 || number == 5 || number == 7 {
		return true
	}
	if number%2 == 0 || number%3 == 0 || number%5 == 0 {
		return false
	}
	sqrt := int(math.Sqrt(float64(number)))
	for i := 5; i <= sqrt; i += 6 {
		if number%i == 0 || number%(i+2) == 0 {
			return false
		}
	}
	return true
}

func delDigitLeftOrRight(number int) []int {
	result := []int{number}
	numberStr := fmt.Sprintf("%d", number)
	power := len(numberStr) - 1
	for k := 1; k <= power; k++ {
		left := numberStr[k:]
		right := numberStr[:len(numberStr)-k]
		if left != "" {
			l, _ := fmt.Sscan(left, &l)
			result = append(result, l)
		}
		if right != "" {
			r, _ := fmt.Sscan(right, &r)
			result = append(result, r)
		}
	}
	return result
}

func euler37() {
	n := 11

	startTime := time.Now()
	dimension := 1000000
	sieve := eratosfen(dimension)
	primes := make([]int, 0)
	for i := 2; i <= dimension; i++ {
		if sieve[i] && i != 2 && i != 3 && i != 5 && i != 7 {
			primes = append(primes, i)
		}
	}
	trucalablePrimes := make([]int, 0)
	counter := 0
	for len(trucalablePrimes) < n && counter < len(primes) {
		nextPrime := primes[counter]
		counter++
		allPrime := true
		for _, v := range delDigitLeftOrRight(nextPrime) {
			if !isPrime(v) {
				allPrime = false
				break
			}
		}
		if allPrime {
			trucalablePrimes = append(trucalablePrimes, nextPrime)
		}
	}
	sum := 0
	for _, v := range trucalablePrimes {
		sum += v
	}
	fmt.Println(sum)
	fmt.Println(time.Since(startTime))

	startTime = time.Now()
	c := 0
	s := 0
	i := n
	for c < n {
		allPrime := true
		for _, v := range delDigitLeftOrRight(i) {
			if !isPrime(v) {
				allPrime = false
				break
			}
		}
		if allPrime {
			c++
			s += i
		}
		i += 2
	}
	fmt.Println(s)
	fmt.Println(time.Since(startTime))
}

func main() {
	euler37()
}

File: python/euler37.py
# Найдите сумму единственных одиннадцати простых чисел, из которых можно
# выбрасывать цифры как справа налево, так и слева направо, но числа при
# этом остаются простыми.

from datetime import datetime
from functools import lru_cache
from typing import Generator, List


def eratosfen(number: int) -> List[int]:
    """
    Решето Эратосфена: генерация простых чисел вплоть до заданного числа.

    Args:
        number: Верхняя граница диапазона.

    Returns:
        Список простых чисел.
    """
    sieve = list(range(number + 1))
    sieve[1] = 0
    for i in sieve:
        if i > 1:
            for j in range(2 * i, len(sieve), i):
                sieve[j] = 0
    return sieve


@lru_cache(2 ** 5)
def isprime(number: int) -> bool:
    """Проверка числа на простоту перебором делителей.

    Args:
        number: Число для проверки.

    Returns:
        True если число простое, иначе False.
    """
    if number in {2, 3, 5, 7}:
        return True
    if number < 2 or number % 2 == 0:
        return False
    if number % 3 == 0 or number % 5 == 0:
        return False
    a, k = number, 0
    for i in range(5, int(number ** 0.5), 6):
        k += 1 if a % i == 0 or a % (i + 2) == 0 else 0
    return k <= 0


def del_digit_left_or_right(number: int) -> Generator[int, None, None]:
    """Возвращает число без правой или левой цифры поочередно.

    Args:
        number: Исходное число.

    Yields:
        Число с удаленными цифрами.
    """
    yield number
    number_str = str(number)
    for k in range(1, len(number_str)):
        yield int(number_str[k:])
        yield int(number_str[:-k])


def euler37() -> None:
    """
    Решение задачи Эйлера №37.

    Найдите сумму одиннадцати простых чисел, из которых можно
    выбрасывать цифры как справа налево, так и слева направо.
    """
    n = 11

    # 1 вариант
    start_time = datetime.now()
    dimension = 1_000_000
    primes = list(set(eratosfen(dimension)) - {0, 2, 3, 5, 7})
    trucalable_primes, counter = [], 0
    while len(trucalable_primes) < n or primes[counter] > dimension:
        next_prime, counter = primes[counter], counter + 1
        if all(map(isprime, del_digit_left_or_right(next_prime))):
            trucalable_primes.append(next_prime)
    print(sum(trucalable_primes))
    print(datetime.now() - start_time)

    # 2 вариант
    start_time = datetime.now()
    c, s = 0, 0
    i = n
    while c < n:
        if all(map(isprime, del_digit_left_or_right(i))):
            c += 1
            s += i
        i += 2
    print(s)
    print(datetime.now() - start_time)


if __name__ == "__main__":
    euler37()

